### PR TITLE
[jit] Support MultiheadedAttention module

### PR DESCRIPTION
**Pull request by driazati**
Stacked PRs
 * #28561 - [jit] Make `nn.Transformer` TorchScript compatible
 * **#28555 - [jit] Support MultiheadedAttention module**

This makes MultiheadedAttention TorchScript compatible

It also breaks BC-compatibility for old models that do not have `_qkv_same_embed_dim` as an attribute.

Differential Revision: [D18124746](https://our.internmc.facebook.com/intern/diff/18124746/)

### CODE REVIEW

**Comment by suo** [comment]
This looks fine to me but it seems the only controversial part of it is the `__setstate__`/bc business. I'll let @zhangguanheng66 review and accept

File: torch/nn/modules/activation.py [hunk 1]
```diff
@@ -756,30 +773,25 @@ def forward(self, query, key, value, key_padding_mask=None,
         - attn_output_weights: :math:`(N, L, S)` where N is the batch size,
           L is the target sequence length, S is the source sequence length.
         """
-        if hasattr(self, '_qkv_same_embed_dim') and self._qkv_same_embed_dim is False:
+        if not self._qkv_same_embed_dim:
             return F.multi_head_attention_forward(
                 query, key, value, self.embed_dim, self.num_heads,
-                None, self.in_proj_bias,  # set self.in_proj_weight = None
+                self.in_proj_weight, self.in_proj_bias,
```
Comment (zhangguanheng66) [inline torch/nn/modules/activation.py]: If `self._qkv_same_embed_dim` is False,  you don't have `self.in_proj_weight` here, right?

Comment (driazati) [reply torch/nn/modules/activation.py]: The latest commit adds it as a `None` valued parameter

File: torch/nn/modules/activation.py [hunk 2]
```diff
@@ -691,8 +696,12 @@ def __init__(self, embed_dim, num_heads, dropout=0., bias=True, add_bias_kv=Fals
             self.q_proj_weight = Parameter(torch.Tensor(embed_dim, embed_dim))
             self.k_proj_weight = Parameter(torch.Tensor(embed_dim, self.kdim))
             self.v_proj_weight = Parameter(torch.Tensor(embed_dim, self.vdim))
+            self.register_parameter('in_proj_weight', None)
         else:
             self.in_proj_weight = Parameter(torch.empty(3 * embed_dim, embed_dim))
+            self.register_parameter('q_proj_weight', None)
```
Comment (zhangguanheng66) [inline torch/nn/modules/activation.py]: why we want to keep those None parameters around? Is this required by JIT?

Comment (driazati) [reply torch/nn/modules/activation.py]: Yeah, the jit requires everything referenced on `self` in `forward` to be defined (even if it is define to be `None`). So even though the code will not execute it still needs to compile, hence the `None` parameters so the parameter definitions are the same for both paths (this is also the reason their names must be in `__constants__`)

Comment (zhangguanheng66) [reply torch/nn/modules/activation.py]: Let me ask the DDP people because they don't like unused parameters in general.

Comment (driazati) [reply torch/nn/modules/activation.py]: For more data points this `register_parameter(..., None)` is used in many other modules including all the `Conv` modules, `BatchNorm`, `LayerNorm`, etc.

**Review by zhangguanheng66 — COMMENTED** [review]
LGTM. Let's wait @pietern @mrshenli and see if he has any comments for the unused parameters there.

**Comment by zhangguanheng66** [comment]
Add BC breaking tag for release note.

**Comment by mrshenli** [comment]
Chatted with @zhangguanheng66 on this offline. DDP should be able to handle this as long all processes are initializing the model in the same way, i.e., you cannot have one process setting `self._qkv_same_embed_dim` to `True` and another setting it to `False`. And after passing the model to DDP, the application should never flip that flag and create param tensors which are not there during construction.

**Comment by gchanan** [comment]
I'm thinking something like this: https://github.com/pytorch/pytorch/blob/42faf961c83d192933e199f8228909cc171765ca/torch/nn/modules/batchnorm.py#L85-L98

**Comment by zhangguanheng66** [comment]
> I'm thinking something like this:
> 
> https://github.com/pytorch/pytorch/blob/42faf961c83d192933e199f8228909cc171765ca/torch/nn/modules/batchnorm.py#L85-L98

https://github.com/pytorch/pytorch/pull/24404

**Comment by zhangguanheng66** [comment]
Land https://github.com/pytorch/pytorch/pull/29001 to avoid BC breaking. Users should still use `torch.load` to the old model trained by pytorch v.1.1
